### PR TITLE
add cpp sol for ciel and duel (silver)

### PR DESCRIPTION
**Pull request by ryanchou-dev**
;w;

---
_If any of the below doesn't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which includes the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making them.

### CODE REVIEW

**Comment by ryanchou-dev** [comment]
let me know if i got the time complexity wrong

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]
ok

**Review by envyaims — APPROVED** [review]
nice :)

File: solutions/cf-321B.mdx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+In the first approach, note that since we're not looking towards defeating all of the cards, we gain **nothing** from attacking Jiro's defense cards. Thus, we'll pair up Jiro's weakest attack cards against our strongest attack cards.
+
+In the second approach, to make sure that we don't lose any attack damage, we'll pair up Jiro's cards with the **smallest** card greater than it.
+
+At the end, we'll return the larger quantity.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(NM \log NM)$
```
Comment (maggie-j-liu) [inline solutions/cf-321B.mdx:26]: wait where does the $\log NM$ come from?

Comment (ryanchou-dev) [reply solutions/cf-321B.mdx:26]: sorting

wait i got it wrong

**Review by maggie-j-liu — APPROVED** [review]

File: solutions/cf-321B.mdx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+```cpp
+#include <iostream>
+#include <algorithm>
+#include <string>
+#include <vector>
+
+using std::vector;
+using std::string;
+using std::endl;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:39]:
```suggestion
using namespace std;
```

File: solutions/cf-321B.mdx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+**Time Complexity:** $\mathcal{O}(NM)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+#include <algorithm>
+#include <string>
+#include <vector>
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:35]: Maybe use `#include <bits/stdc++.h>`?

File: solutions/cf-321B.mdx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+struct Card {
+	// 1 -> ATK | 0 -> DEF
+	bool type;
+	int strength;
+};
+
+int main() {
+	int n, m;
+	std::cin >> n >> m;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:49]:
```suggestion
	cin >> n >> m;
```

File: solutions/cf-321B.mdx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+	vector<Card> jiro;
+	vector<Card> ciel;
+
+	for (int i = 0; i < n; i++) {
+		string type;
+		int strength;
+		std::cin >> type >> strength;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:57]:
```suggestion
		cin >> type >> strength;
```

File: solutions/cf-321B.mdx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+		jiro.push_back({type == "ATK", strength});
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:59]: Explain this maybe

File: solutions/cf-321B.mdx [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	}
+
+	for (int i = 0; i < m; i++) {
+		int strength;
+		std::cin >> strength;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:64]:
```suggestion
		cin >> strength;
```

File: solutions/cf-321B.mdx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+		ciel.push_back({1, strength});
+	}
+
+	// first case: only hit ATK + assign weak cards with strong cards
+
+	// sort by type and then by strength
+	std::sort(jiro.begin(), jiro.end(), [](const Card &a, const Card &b) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:72]:
```suggestion
	sort(jiro.begin(), jiro.end(), [](const Card &a, const Card &b) {
```

File: solutions/cf-321B.mdx [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		if (a.type == b.type) {
+			return a.strength < b.strength;
+		}
+		return a.type > b.type; 
+	});
+
+	std::sort(ciel.begin(), ciel.end(), [](const Card &a, const Card &b) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:79]:
```suggestion
	sort(ciel.begin(), ciel.end(), [](const Card &a, const Card &b) {
```

File: solutions/cf-321B.mdx [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+		// break if either we're attacking DEF cards or we don't gain anything
+		if (!jiro[i].type || ciel[i].strength <= jiro[i].strength){
+			break;
+		}
+
+		appr1 += ciel[i].strength - jiro[i].strength;
+	}
+
+	std::reverse(jiro.begin(), jiro.end());
+	std::reverse(ciel.begin(), ciel.end());
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:94]:
```suggestion
	reverse(jiro.begin(), jiro.end());
	reverse(ciel.begin(), ciel.end());
```

File: solutions/cf-321B.mdx [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		return a.type > b.type; 
+	});
+
+	std::sort(ciel.begin(), ciel.end(), [](const Card &a, const Card &b) {
+		return a.strength > b.strength;
+	});
+
+	int appr1 = 0;
+	for (int i = 0; i < std::min(n, m); i++) {
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:84]:
```suggestion
	for (int i = 0; i < min(n, m); i++) {
```

File: solutions/cf-321B.mdx [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+
+	// add up all of the direct damage done
+	for (int i = 0; i < m; i++) {
+		if (!done[i]) {
+			appr2 += ciel[i].strength;
+		}
+	}
+
+	std::cout << std::max(appr2, appr1) << endl;
```
Comment (Aadit-Ambadkar) [inline solutions/cf-321B.mdx:134]:
```suggestion
	cout << max(appr2, appr1) << "\n";
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]
Watch all of this get rejected.

File: solutions/cf-321B.mdx (hunk 2 above)
Comment (maggie-j-liu) [reply solutions/cf-321B.mdx:39]: we don't really regulate coding style unless it affects readability, which this (and the headers) doesn't

File: solutions/cf-321B.mdx (hunk 6 above)
Comment (ryanchou-dev) [reply solutions/cf-321B.mdx:59]: it's just input?

**Comment by ryanchou-dev** [comment]
err yeah, all of these suggestions are just coding style. thanks for reviewing though!

**Review by SansPapyrus683 — APPROVED** [review]

File: solutions/cf-321B.mdx (hunk 6 above)
Comment (ryanchou-dev) [reply solutions/cf-321B.mdx:59]: feel free to unresolve if you have a diff opinion

**Comment by SansPapyrus683** [comment]
> Watch all of this get rejected.

literally 1984

**Pull request merged**